File: PartitionInt.py
//Array Partition for Integers in python
def quickSort(alist):
   quickSortHelper(alist,0,len(alist)-1)

def quickSortHelper(alist,first,last):
   if first<last:

       pivotValue = partition(alist,first,last)

       quickSortHelper(alist,first,pivotValue-1)
       quickSortHelper(alist,pivotValue+1,last)


def partition(alist,first,last):
   pivotvalue = alist[first]

   leftmark = first+1
   rightmark = last

   done = False
   while not done:

       while leftmark <= rightmark and \
               alist[leftmark] <= pivotvalue:
           leftmark = leftmark + 1

       while alist[rightmark] >= pivotvalue and \
               rightmark >= leftmark:
           rightmark = rightmark -1

       if rightmark < leftmark:
           done = True
       else:
           temp = alist[leftmark]
           alist[leftmark] = alist[rightmark]
           alist[rightmark] = temp

   
   [alist[first], alist[rightmark]]= [alist[rightmark], alist[first]]
   

   return rightmark

alist = [3,2,1,7,6,8,0,9,4]
print("Before Sort")
print(alist)
quickSort(alist)
print("After Sort")
print(alist)

File: PartitionInt.cpp
//Array Partition for Integers in C++
#include <iostream>
 
void quickSort(int a[], int first, int last);
int pivot(int a[], int first, int last);
void swap(int& a, int& b);
void swapNoTemp(int& a, int& b);
void print(int array[], const int& N);
 
using namespace std;
 
int main()
{
    int test[] = {3,2,1,7,6,8,0,9,4};
    int N = sizeof(test)/sizeof(int);
 
    cout << "array length :"  << N << endl;
 
    cout << "Before sorting : " << endl;
    print(test, N);
 
    quickSort(test, 0, N-1);
 
    cout << endl << endl << "After sorting : " << endl;
    print(test, N);
     
    return 0;
}
 
//QuickSort Function
void quickSort( int a[], int first, int last ) 
{
    int pivotValue;
 
    if(first < last)
    {
        pivotValue = pivot(a, first, last);
        quickSort(a, first, pivotValue-1);
        quickSort(a, pivotValue+1, last);
    }
}
 
//Select Pivot Point
int pivot(int a[], int first, int last) 
{
    int  p = first;
    int pivotValue = a[first];   
    for(int i = first+1 ; i <= last ; i++)
    {
        if(a[i] <= pivotValue)
        {
            p++;
            swap(a[i], a[p]);
        }
    }
 
    swap(a[p], a[first]);
 
    return p;
}
  
//Swapping Values
void swap(int& a, int& b)
{
    int temp = a;
    a = b;
    b = temp;
}
 
void print(int a[], const int& N)
{
    for(int i = 0 ; i < N ; i++)
        cout << "array[" << i << "] = " << a[i] << endl;
}

